rfd/0005-kubernetes-service.md

```diff
@@ -0,0 +1,302 @@
+---
+authors: Andrew Lytvynov (andrew@gravitational.com)
+state: draft
+---
+
+# RFD 5 - Kubernetes (k8s) service enhancements
+
+## What
+
+A series of enhancements to the k8s integration:
+- k8s service separated from Proxy service
+- support for multiple k8s clusters per Teleport cluster
+- support for k8s clusters with blocked ingress (without Trusted
+  Clusters)
+- k8s cluster info available via Teleport API
+- automatic login prompt in `kubectl` on session expiry
+- audit logging of all k8s API requests
+- RBAC for k8s cluster access
+
+## Why
+
+k8s integration in Teleport has been implemented as an optional addon to
+the Proxy service. It has several limitations:
+
+- only one k8s cluster per Teleport cluster supported
+  - multiple clusters are supported via Trusted Clusters, which is more
+    operationally complex
+  - a k8s cluster behind network obstacles (firewalls/NATs/etc) also
+    requires a Trusted Cluster
+- doesn't follow the "separation of concerns" principle in the proxy
+- expired user certificate leads to obscure TLS errors from `kubectl`
+- no easy way to tell whether a Teleport cluster has a k8s cluster
+  behind it
+- audit log ignores all non-interactive k8s requests
+
+## Details
+
+### Backwards compatibility
+
+The old k8s integration remains supported and unchanged. Users who don't
+migrate to the new integration will not be broken.
+
+The new k8s service will be preferred to the old Proxy-based integration
+and all new features will be added to the new service only.
+
+### Service definition
+
+To separate k8s integration from Proxy (and all associated semantics,
+like the discovery protocol), a new `kubernetes_service` will be added in
+`teleport.yaml` and can be enabled independently from `proxy_service`. This is
+similar to `ssh_service` and the upcoming `application_service`.
+
+The new k8s service will inherit all of the configuration and behavior
+of the existing `kubernetes` section of `proxy_service`:
+
+```yaml
+# Old format:
+proxy_service:
+    kubernetes:
+        enabled: yes
+        public_addr: [k8s.example.com:3026]
+        listen_addr: 0.0.0.0:3026
+        kubeconfig_file: /secrets/kubeconfig
+```
+
+but will exist in a new top-level config:
+
+```yaml
+# New format:
+kubernetes_service:
+    enabled: yes
+    public_addr: [k8s.example.com:3026]
```

Comment:
With this new service? Do customers still need to setup TLS for this endpoint? Or are they using something internal from Kubernetes. Does the FQDN represent anything else in that case?

Reply:
Resolved on the review call.